radixdlt-java/src/main/java/com/radixdlt/client/core/ledger/InMemoryAtomStore.java

```diff
@@ -43,102 +60,131 @@
-
+	public Observable<Long> onSync(RadixAddress address) {
+		return Observable.create(emitter -> {
+			synchronized (lock) {
+				if (syncedMap.getOrDefault(address, false)) {
+					emitter.onNext(System.currentTimeMillis());
+				}
+
+				final CopyOnWriteArrayList<ObservableEmitter<Long>> syncers;
+				if (!allSyncers.containsKey(address)) {
+					syncers = new CopyOnWriteArrayList<>();
+					allSyncers.put(address, syncers);
+				} else {
+					syncers = allSyncers.get(address);
+				}
+				syncers.add(emitter);
+				emitter.setCancellable(() -> syncers.remove(emitter));
+			}
+		});
+	}
 
-						final boolean include;
-						if (lastUpdate == null) {
-							include = nextUpdate.getType() == Type.STORE;
+	@Override
+	public Stream<Particle> getUpParticles(RadixAddress address) {
+		synchronized (lock) {
+			return particleIndex.entrySet().stream()
+				.filter(e -> e.getKey().getShardables().contains(address)
+					&& e.getValue().stream().flatMap(a -> {
+						AtomObservation observation = atoms.get(a);
+						if (observation.isStore()) {
+							return observation.getAtom().spunParticles()
+								.filter(s -> s.getParticle().equals(e.getKey()))
+								.map(SpunParticle::getSpin);
 						} else {
-							// Soft observation should not be able to update a hard state
-							// Only update if type changes
-							include = (!nextUpdate.isSoft() || lastUpdate.isSoft())
-								&& nextUpdate.getType() != lastUpdate.getType();
-						}
-
-						// Core currently does not guarantee DELETEs being emitted in order so need this hack for now
-						// That is, create soft deletes for any known missing deletes from dependent atoms
-						List<AtomObservation> missingDeletes = new ArrayList<>();
-						if (nextUpdate.getType() == Type.DELETE && include) {
-							// TODO: Should we DELETE atoms which are dependent on these atoms as well?
-							List<AtomObservation> observationsContainingDown = observation.getAtom().particles(Spin.UP)
-								.flatMap(p -> atomsObservationState.atomContainingDown(p)
-									.map(o -> o.isStore() && !o.getAtom().getHid().equals(observation.getAtom().getHid())
-										? Stream.of(o) : Stream.<AtomObservation>empty())
-									.orElse(Stream.empty())
-								)
-								.collect(Collectors.toList());
-
-							if (!observationsContainingDown.isEmpty()) {
-								observationsContainingDown.forEach(o -> missingDeletes.add(AtomObservation.softDeleted(o.getAtom())));
-							}
-						}
-
-						// Should always update observation state if going from soft to hard observation
-						final boolean isSoftToHard = lastUpdate != null && lastUpdate.isSoft() && !nextUpdate.isSoft();
-
-						if (include || isSoftToHard) {
-							atomsObservationState.put(observation);
-							missingDeletes.forEach(atomsObservationState::put);
+							return Stream.empty();
 						}
+					}).reduce(Spin.NEUTRAL, (s0, s1) -> SpinStateMachine.isAfter(s0, s1) ? s0 : s1) == Spin.UP
```

Comment:
Could make some performance optimisations here given the fact that we only need to know that it has not been DOWNed, not really critical though

Reply:
Performance isn't concern right now but correctness. This class is going to change a lot to get the right modules in the right place.